### PR TITLE
Classifier Reorganization

### PR DESCRIPTION
**Pull request by ctrl-z-9000-times**
For issue #335.

I redesigned the API.  The algorithms are mostly unchanged.

Sorry, I did not follow my own advice: this is one big PR.  A lot of things changed and so it may be easier to review this as new code, instead of looking at the diffs.

Tests:
I ported the original python tests for the SDRClassifier and updated them to use the new API.  I updated most of the C++ tests to use the new API as well (deleted 1 redundant C++ test).

Outstanding Tasks:
- [x] Serialization
- [x] Python Docs
- [x] C++ example usage & unit test for it
- [x] Python example usage & unit test for it
- [x] Default values for alpha - Set to 0.001 as it was in Numenta's nupic.
- [x] Document in API_CHANGELOG
- [x] Fix windows build warnings
- [x] Docs use word category, explain buckets once in class level doc

### CODE REVIEW

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 1]
```diff
@@ -21,489 +21,211 @@
...
+      err_msg << "), now given SDR with dimensions ( ";
+      for( auto dim : pattern.dimensions )
+        { err_msg << dim << " "; }
+      err_msg << ").";
+      NTA_THROW << err_msg.str();
+  }
+
+  // Accumulate feed forward input.
+  PDF probabilities( numCategories_, 0.0f );
+  for( const auto& bit : pattern.getSparse() ) {
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: nit, auto, not auto& for primitive types

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 2]
```diff
@@ -21,489 +21,211 @@
...
-        }
-      }
-    }
-  }
+PDF Classifier::infer(const SDR & pattern)
+{
+  // Check input dimensions, or if this is the first time the Classifier has
+  // been used then initialize it with the given SDR's dimensions.
+  if( dimensions_.empty() ) {
+    dimensions_ = pattern.dimensions;
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: dimensions cannot change during runtime, let's set them during init (will be able to move this whole if-elseif into initialize)

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 3]
```diff
@@ -21,489 +21,211 @@
...
-      // if doing 0-step ahead prediction, we shouldn't use any
-      // knowledge of the classification input during inference
-      if( steps_.at(0) == 0 ) {
-        actValueVector.push_back( 0.0f );
-      }
-      else {
-        actValueVector.push_back( actValue[0] );
+  // Check if this is a new category & resize the weights table to hold it.
+  const auto maxCategoryIdx = *max_element(categoryIdxList.begin(), categoryIdxList.end());
+  if( maxCategoryIdx >= numCategories_ ) {
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: - same as in infer, can you move this all into initialize? It won't change during lifetime of the classifier. 
- Also, may need some doc on categoryIdxList
  - can we pass the encoder here, to obtain both values?

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 4]
```diff
@@ -21,489 +21,211 @@
...
+  for( size_t i = 0; i < likelihoods.size(); i++ ) {
     likelihoods[i] = targetDistribution[i] - likelihoods[i];
   }
   return likelihoods;
 }
 
 
-void SDRClassifier::softmax_(const vector<Real64>::iterator begin,
-                             const vector<Real64>::iterator end) {
+void nupic::algorithms::sdr_classifier::softmax(PDF::iterator begin, PDF::iterator end) {
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: nit, why not `using namespace nupic::algorithms::sdr_classifier` ?

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 5]
```diff
@@ -21,489 +21,211 @@
...
-    outStream << pattern.size() << " ";
-    for (auto &pattern_j : pattern) {
-      outStream << pattern_j << " ";
-    }
+void Predictor::initialize(const vector<UInt> &steps, Real alpha)
+{
+  steps_ = steps;
+  if( steps_.empty() ) {
+    // steps_.push_back( 0u );
+    NTA_THROW << "Required argument steps is empty!";
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: nit, NTA_CHECK for consistency

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 6]
```diff
@@ -21,489 +21,211 @@
...
-      for (UInt j = 0; j < maxBucketIdx_; j++) {
-	Real64 val;
-        inStream >> val;
-	if(val == 0.0) continue; //load sparse map only
-	m[i][j] = val;
-      }
-    }
-    weightMatrix_[step] = m;
-  }
+Predictions Predictor::infer(UInt recordNum, const SDR &pattern)
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: nit, `const recordNum`

File: src/nupic/algorithms/SDRClassifier.hpp [hunk 7]
```diff
@@ -22,203 +22,213 @@
...
  */
-const Int ACTUAL_VALUES = -1;
-using ClassifierResult = std::map<Int, PDF>;
+UInt getClassification( const PDF & data );
 
 
-class SDRClassifier : public Serializable
+/**
+ * The SDR Classifier takes the form of a single layer classification network
+ * that takes SDRs as input and outputs a predicted distribution of classes.
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.hpp]: high-level: this whole classifier is not a part of HTM, can we find a simple 3rd party NN library to delegate the task to? And just have a wrapper to call it with SDR and input?

File: src/nupic/algorithms/SDRClassifier.hpp [hunk 8]
```diff
@@ -22,203 +22,213 @@
...
-                       const std::vector<Real64> &actValueList, bool category,
-                       bool learn, bool infer, ClassifierResult &result);
+  PDF infer(const sdr::SDR & pattern);
 
   /**
-   * Returns the class with the greatest probablility.
+   * Learn from example data.
+   *
+   * @param pattern:  The active input bit SDR.
+   * @param categoryIdxList:  The current categories or bucket indices.
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.hpp]: please explain where the categories are obtained from

File: src/test/unit/algorithms/SDRClassifierTest.cpp [hunk 9]
```diff
@@ -37,302 +36,167 @@
...
-  c1.compute(0u, input1, bucketIdxList1, actValueList1, false, true, true, result);
-
-  {
-    stringstream ss;
-    EXPECT_NO_THROW(c1.save(ss));
-    EXPECT_NO_THROW(c2.load(ss));
-  }
-  ASSERT_EQ(c1, c2);
+TEST(SDRClassifierTest, SaveLoad) {
+  ASSERT_EQ(1, 2); // TODO!
```
Comment (breznak) [inline src/test/unit/algorithms/SDRClassifierTest.cpp]: known TODO serialization

**Comment by breznak** [comment]
Also, note in API_CHANGELOG

File: src/examples/mnist/MNIST_SP.cpp [hunk 10]
```diff
@@ -39,16 +39,16 @@ using namespace std;
 using namespace nupic;
 
 using nupic::algorithms::spatial_pooler::SpatialPooler;
-using nupic::algorithms::sdr_classifier::SDRClassifier;
-using nupic::algorithms::sdr_classifier::ClassifierResult;
+using nupic::algorithms::sdr_classifier::Classifier;
```
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply src/examples/mnist/MNIST_SP.cpp]: I wanted to change the classes name because the API is incompatible.

File: src/examples/mnist/MNIST_SP.cpp [hunk 11]
```diff
@@ -148,16 +137,8 @@ void test() {
...
-    clsr.compute(sp.getIterationNum(), columns.getSparse(),
-      /* bucketIdxList */   {},
-      /* actValueList */    {},
-      /* category */        true,
-      /* learn */           false,
-      /* infer */           true,
-                            result);
     // Check results
-    if(clsr.getClassification( result[0] ) == label)
+    if( getClassification( clsr.infer( columns ) ) == label)
```
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply src/examples/mnist/MNIST_SP.cpp]: Both `Classifier` and `Predictor` use this function, so I don't want it bound to one class or the other.

Also, can we change this functions name to `argmax`?

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 12]
```diff
@@ -21,489 +21,211 @@
...
+PDF Classifier::infer(const SDR & pattern)
+{
+  // Check input dimensions, or if this is the first time the Classifier has
+  // been used then initialize it with the given SDR's dimensions.
+  if( dimensions_.empty() ) {
+    dimensions_ = pattern.dimensions;
+    while( weights_.size() < pattern.size ) {
+      weights_.push_back( vector<Real>( numCategories_, 0.0f ));
+    }
+  } else if( pattern.dimensions != dimensions_ ) {
```
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/SDRClassifier.cpp]: I'd like for this check to always run.  Otherwise its liable to segfault if the user uses the wrong SDR.

File: src/nupic/algorithms/SDRClassifier.cpp (hunk 2 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/SDRClassifier.cpp]: This way the user doesn't need to tell us the dimensions ahead of time.  It's one less argument to remember.

File: src/nupic/algorithms/SDRClassifier.cpp (hunk 3 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/SDRClassifier.cpp]: > same as in infer, can you move this all into initialize? It won't change during lifetime of the classifier.

I think it's easier for the user to not need to initialize this.  One less argument to remember.

Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/SDRClassifier.cpp]: > Also, may need some doc on categoryIdxList

Yes.
In this PR I use the words: "categoryIdx" and bucketIdx" interchangeably (they're the same thing in this context).  I'd like to use only categoryIdx and provide an example of converting an integer into a bucketIdx.

File: src/nupic/algorithms/SDRClassifier.hpp (hunk 7 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/SDRClassifier.hpp]: I'd rather not:
- This was already implemented & tested by numenta.
- This works pretty well. I don't think a NN library would do significantly better than this.
- This is a very simple network, I think a whole NN framework is overkill.

File: src/examples/mnist/MNIST_SP.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (breznak) [reply src/examples/mnist/MNIST_SP.cpp]: not sure if that'll be more, or less confusing. but no big deal

File: src/nupic/algorithms/SDRClassifier.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (breznak) [reply src/nupic/algorithms/SDRClassifier.cpp]: the NTA_CHECK, but not a high priority, so as you want

File: src/nupic/algorithms/SDRClassifier.hpp (hunk 7 above)
Comment (breznak) [reply src/nupic/algorithms/SDRClassifier.hpp]: > This works pretty well. I don't think a NN library would do significantly better than this. 

well, this is a room for experimentation. But ok, not concern of this PR

File: src/nupic/algorithms/SDRClassifier.cpp (hunk 4 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/SDRClassifier.cpp]: I do have the line `using namespace nupic::algorithms::sdr_classifier` at the top of the file, but for some unknown reason I needed to make the namespace explicit for the compiler to find it.

File: src/nupic/algorithms/SDRClassifier.hpp [hunk 13]
```diff
@@ -34,7 +34,11 @@
 
 #include <nupic/types/Types.hpp>
 #include <nupic/types/Sdr.hpp>
+
 #include <nupic/types/Serializable.hpp>
+#include <cereal/types/vector.hpp>
+#include <cereal/types/deque.hpp>
+#include <cereal/types/map.hpp>
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.hpp]: make these included in Serializable (vector already is. does cereal.hpp not import all of them?), so the implementing class does not have to do anything than Serializable.hpp

File: src/nupic/types/Serializable.hpp [hunk 14]
```diff
@@ -217,8 +217,8 @@ class Serializable {
 	//       Remove the following two lines.
 
   // These must be implemented by the subclass.
-  virtual void save(std::ostream &stream) const {};
-  virtual void load(std::istream &stream) {};
+  virtual void save(std::ostream &stream) const { saveToStream_ar(stream); };
+  virtual void load(std::istream &stream) { loadFromStream_ar(stream); };
```
Comment (breznak) [inline src/nupic/types/Serializable.hpp]: what is this global change? It probably should not be part of this PR

File: src/test/unit/algorithms/SDRClassifierTest.cpp [hunk 15]
```diff
@@ -150,30 +150,31 @@ TEST(ClassifierTest, MultipleCategories) {
...
-  // }
-  // ASSERT_EQ(c1, c2);
-
-  // ClassifierResult result1, result2;
-  // c1.compute(1u, input1, bucketIdxList1, actValueList1, false, true, true, result1);
-  // c2.compute(1u, input1, bucketIdxList1, actValueList1, false, true, true, result2);
-
-  // ASSERT_EQ(result1[1u], result2[1]);
+  vector<UInt> steps{ 1u };
+  Predictor c1(steps, 0.1f);
```
Comment (breznak) [inline src/test/unit/algorithms/SDRClassifierTest.cpp]: TODO also test for Classifier

File: src/test/unit/algorithms/SDRClassifierTest.cpp [hunk 16]
```diff
@@ -150,30 +150,31 @@ TEST(ClassifierTest, MultipleCategories) {
...
+
+  // Save and load.
+  stringstream ss;
+  EXPECT_NO_THROW(c1.save(ss));
+  Predictor c2;
+  EXPECT_NO_THROW(c2.load(ss));
+
+  // Expect identical results.
+  const auto c2_out = c2.infer( 0u, A );
+  ASSERT_EQ(c1_out, c2_out);
```
Comment (breznak) [inline src/test/unit/algorithms/SDRClassifierTest.cpp]: also check that c1, c2 are equal

File: src/test/unit/algorithms/SDRClassifierTest.cpp (hunk 15 above)
Comment (ctrl-z-9000-times) [reply src/test/unit/algorithms/SDRClassifierTest.cpp]: The `Predictor` contains a `Classifier` so this should test both at once.

File: src/nupic/algorithms/SDRClassifier.hpp (hunk 13 above)
Comment (dkeeney) [reply src/nupic/algorithms/SDRClassifier.hpp]: Yes, those includes should be in Serializable.hpp

File: src/nupic/types/Serializable.hpp (hunk 14 above)
Comment (dkeeney) [reply src/nupic/types/Serializable.hpp]: No.  The objective is to keep save/load as-is until Cereal is complete and then remove them.  That is so that we don't have all serialization broken until we finish.  So, don't redirect to saveToStream_ar( ) and loadFromStream_ar( )

Comment (ctrl-z-9000-times) [reply src/nupic/types/Serializable.hpp]: I made this change so that my tests can call save/load.  Since I rewrote the serialization for these classes, they do not implement old save/load, only cereal.

Save/load are the public facing API, and eventually the plan is to  rename "saveToStream_ar" to "save".  This change renames them except for the classes which still implement the old methods.

File: src/nupic/algorithms/SDRClassifier.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/SDRClassifier.cpp]: I added the extra verbose output was in response to a specific and frustrating error.  The following is incorrect but will compile:
```
Classifier C();
vector<UInt> sparse({ 4,5,6 });
C.infer( sparse );
```
Instead of failing because the input `sparse` is not an SDR, the C++ compiler creates an SDR with dimensions [4,5,6].

File: src/nupic/types/Serializable.hpp (hunk 14 above)
Comment (dkeeney) [reply src/nupic/types/Serializable.hpp]: The old implementation of save/load must remain in place until Cerial is complete.  They don't mix correctly with save_ar and load_ar.

Comment (dkeeney) [reply src/nupic/types/Serializable.hpp]: If you have a class that does not have the original save/load then just don't implement them.

File: src/nupic/algorithms/SDRClassifier.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (breznak) [reply src/nupic/algorithms/SDRClassifier.cpp]: ok, that is a really cunning of the compiler! Good catch

File: src/test/unit/algorithms/SDRClassifierTest.cpp (hunk 16 above)
Comment (ctrl-z-9000-times) [reply src/test/unit/algorithms/SDRClassifierTest.cpp]: I can't check that.  I removed `operator==` because this is the only place where it would be applicable.  If both the original and serialized/deserialized class instances yield the same results then for practical purposes they're identical.

**Comment by ctrl-z-9000-times** [comment]
Cereal is implemented and tested.  See sdrclassifier.hpp lines 130 - 140.

**Comment by dkeeney** [comment]
Oh,  somehow I missed them.  Nicely done too :)

**Comment by ctrl-z-9000-times** [comment]
Thanks for the review!

File: src/nupic/algorithms/SDRClassifier.cpp (hunk 3 above)
Comment (breznak) [reply src/nupic/algorithms/SDRClassifier.cpp]: did you get to do these?

Comment (breznak) [reply src/nupic/algorithms/SDRClassifier.cpp]: ..ok, explained here 
https://github.com/htm-community/nupic.cpp/pull/401/files#diff-d6a5469efcd8e0a1cba95db1b5906037R81
